tooling/nargo/src/foreign_calls/mocker.rs

```diff
@@ -0,0 +1,176 @@
+    /// Optionally match the parameters
+    params: Option<Vec<ForeignCallParam<F>>>,
+    /// The parameters with which the mock was last called
+    last_called_params: Option<Vec<ForeignCallParam<F>>>,
+    /// The result to return when this mock is called
+    result: ForeignCallResult<F>,
+    /// How many times should this mock be called before it is removed
+    times_left: Option<u64>,
+}
+
+impl<F> MockedCall<F> {
+    fn new(id: usize, name: String) -> Self {
+        Self {
+            id,
+            name,
+            params: None,
+            last_called_params: None,
+            result: ForeignCallResult { values: vec![] },
+            times_left: None,
+        }
+    }
+}
+
+impl<F: PartialEq> MockedCall<F> {
+    fn matches(&self, name: &str, params: &[ForeignCallParam<F>]) -> bool {
+        self.name == name && (self.params.is_none() || self.params.as_deref() == Some(params))
+    }
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct MockForeignCallExecutor<F> {
+    /// Mocks have unique ids used to identify them in Noir, allowing to update or remove them.
+    last_mock_id: usize,
+    /// The registered mocks
+    mocked_responses: Vec<MockedCall<F>>,
+}
+
+impl<F: AcirField> MockForeignCallExecutor<F> {
+    fn extract_mock_id(
+        foreign_call_inputs: &[ForeignCallParam<F>],
+    ) -> Result<(usize, &[ForeignCallParam<F>]), ForeignCallError> {
+        let (id, params) =
+            foreign_call_inputs.split_first().ok_or(ForeignCallError::MissingForeignCallInputs)?;
+        let id =
+            usize::try_from(id.unwrap_field().try_to_u64().expect("value does not fit into u64"))
+                .expect("value does not fit into usize");
+        Ok((id, params))
+    }
+
+    fn find_mock_by_id(&self, id: usize) -> Option<&MockedCall<F>> {
+        self.mocked_responses.iter().find(|response| response.id == id)
+    }
+
+    fn find_mock_by_id_mut(&mut self, id: usize) -> Option<&mut MockedCall<F>> {
+        self.mocked_responses.iter_mut().find(|response| response.id == id)
+    }
+
+    fn parse_string(param: &ForeignCallParam<F>) -> String {
+        let fields: Vec<_> = param.fields().to_vec();
+        decode_string_value(&fields)
+    }
+}
+
+impl<F: AcirField + Serialize + for<'a> Deserialize<'a>> ForeignCallExecutor<F>
+    for MockForeignCallExecutor<F>
+{
+    fn execute(
+        &mut self,
+        foreign_call: &ForeignCallWaitInfo<F>,
+    ) -> Result<ForeignCallResult<F>, ForeignCallError> {
+        let foreign_call_name = foreign_call.function.as_str();
+        match ForeignCall::lookup(foreign_call_name) {
+            Some(ForeignCall::CreateMock) => {
+                let mock_oracle_name = Self::parse_string(&foreign_call.inputs[0]);
+                assert!(ForeignCall::lookup(&mock_oracle_name).is_none());
+                let id = self.last_mock_id;
+                self.mocked_responses.push(MockedCall::new(id, mock_oracle_name));
+                self.last_mock_id += 1;
```

Comment:
It's not obvious to me why the mocks aren't stored in a `BTreeMap` indexed by ID. I looks like we won't have repeated IDs; I thought there might be multiple single-use responses lined up for the same ID, but not sure that's a use case.

Reply:
Our testing of mocked oracle calls is woefully incomplete atm so this may not be ideal. Let's handle this in a followup.